Sources/SourceKitLSP/Swift/Diagnostic.swift

```diff
@@ -223,9 +223,17 @@ extension CachedDiagnostic {
 ///
 /// Sourcekitd returns parse diagnostics immediately after edits, but we do not want to clear the
 /// semantic diagnostics until we have semantic level diagnostics from after the edit.
-func mergeDiagnostics(old: [CachedDiagnostic], new: [CachedDiagnostic], stage: DiagnosticStage) -> [CachedDiagnostic] {
+///
+/// However, if fallback arguments are being used, we withhold semantic diagnostics in favor of only
+/// emitting syntactic diagnostics.
+func mergeDiagnostics(
+  old: [CachedDiagnostic],
+  new: [CachedDiagnostic],
+  stage: DiagnosticStage,
+  isFallback: Bool
+) -> [CachedDiagnostic] {
   if stage == .sema {
-    return new
+    return isFallback ? old.filter { $0.stage == .parse } : new
```

Comment:
Even if you're falling back, you want the *new* parse diagnostics, not the old ones (otherwise you just get stuck with the old diagnostics permanently).

Also, I think what we want here is to leave mergeDiagnostics alone so that we always cache the semantic diagnostics regardless of fallback, but then filter them before they get sent back in the response.

Reply:
Ah, so there can be parsing stage diagnostics even when `stage == .sema`?  And OK, that SGTM, will do

Reply:
I tried out your suggestion but it doesn't work in the following case:

old (was fallback): stores semantic and parse diagnostics (only emit parse diagnostics though)
new (no longer fallback): parse case. we keep the old semantic diagnostics around, but they were invalid and ignored, but we don't know that since we don't store that information, so we wrongly emit the old semantic diagnostics + new parse diagnostics.

Reply:
Ah good point. Okay, then I think the only thing left is we want to filter to the new parse diagnostics not the old ones.

Reply:
I just checked that as well, I only see semantic diagnostics for the sema stage, so if I do that we lose all diagnostics. Thus we can need the current behavior.

Reply:
Ah I see.  So when in fallback, the flow is:
* parse diagnostics arrive, we will return new (parse) + old (sema, empty)
* sema diagnostics arrive, we will return old (parse), which should be identical to the last publishDiagnostics

Okay

Reply:
Yep